src/probability_theory/stopping.lean

```diff
@@ -0,0 +1,271 @@
+filtration, stopping time, stochastic process
+
+-/
+
+noncomputable theory
+open_locale classical measure_theory nnreal ennreal topological_space probability_theory
+
+namespace measure_theory
+
+/-- A `filtration` on measurable space `α` with σ-algebra `m` is a monotone
+sequence of of sub-σ-algebras of `m`. -/
+structure filtration (α ι : Type*) [preorder ι] (m : measurable_space α) :=
+(seq : ι → measurable_space α)
+(mono : monotone seq) (le : ∀ i : ι, seq i ≤ m)
+
+instance {α ι : Type*} [preorder ι] {m : measurable_space α} :
+  has_coe_to_fun (filtration α ι m) (λ _, ι → measurable_space α) :=
+⟨λ f, f.seq⟩
+
+instance {α ι : Type*} [preorder ι] {m : measurable_space α} : inhabited (filtration α ι m) :=
+⟨⟨λ i, m, monotone_const, λ i, le_rfl⟩⟩
+
+open topological_space
+
+variables {α β ι : Type*} {m : measurable_space α} [measurable_space β]
+
+section
+
+variables [preorder ι]
+
+lemma measurable_set_of_filtration {f : filtration α ι m} {s : set α} {i : ι}
+  (hs : measurable_set[f i] s) : measurable_set[m] s :=
+f.le i s hs
+
+/-- A measure is σ-finite with respect to filtration if it is σ-finite with respect
+to all the sub-σ-algebra of the filtration. -/
+class sigma_finite_filtration (μ : measure α) (f : filtration α ι m) : Prop :=
+(sigma_finite : ∀ ⦃i : ι⦄, @sigma_finite α (f i) (μ.trim (f.le i)))
+
+instance sigma_finite_of_sigma_finite_filtration (μ : measure α) (f : filtration α ι m)
+  [hf : sigma_finite_filtration μ f] (i : ι) :
+  @sigma_finite α (f i) (μ.trim (f.le i)) :=
+by apply hf.sigma_finite -- can't exact here
+
+/-- A sequence of functions `u` is adapted to a filtration `f` if for all `i`,
+`u i` is `f i`-measurable. -/
+def adapted (f : filtration α ι m) (u : ι → α → β) : Prop :=
+∀ ⦃i : ι⦄, measurable[f i] (u i)
+
+namespace filtration
+
+/-- Given a sequence of functions, the natural filtration is the smallest sequence
+of σ-algebras such that that sequence of functions is measurable with respect to
+the filtration. -/
+def natural (u : ι → α → β) (hum : ∀ i, measurable (u i)) : filtration α ι m :=
+{ seq := λ i, ⨆ j ≤ i, measurable_space.comap (u j) infer_instance,
+  mono := λ i j hij, bsupr_le_bsupr' $ λ k hk, le_trans hk hij,
+  le := λ i, bsupr_le (λ j hj s hs, let ⟨t, ht, ht'⟩ := hs in ht' ▸ hum j ht) }
+
+lemma adapted_natural {u : ι → α → β} (hum : ∀ i, measurable[m] (u i)) :
+  adapted (natural u hum) u :=
+λ i, measurable.le (le_bsupr_of_le i (le_refl i) (le_refl _)) (λ s hs, ⟨s, hs, rfl⟩)
+
+end filtration
+
+variables {μ : measure α} {f : filtration α ι m}
+
+/-- A stopping time with respect to some filtration `f` is a function
+`τ` such that for all `i`, the preimage of `{j | j ≤ i}` along `τ` is measurable
+with respect to `f i`.
+
+Intuitively, the stopping time `τ` describes some stopping rule such that at time
+`i`, we may determine it with the information we have at time `i`. -/
+def is_stopping_time (f : filtration α ι m) (τ : α → ι) :=
+∀ i : ι, measurable_set[f i] $ {x | τ x ≤ i}
+
+section
```

Comment:
Could you give names to your sections? It helps later contributors find the right place for their lemmas.

Reply:
I don't know why I had a section there in the first place :) I I remove the section instead